src/backend/distributed/executor/executor.c

```diff
@@ -1211,9 +1231,11 @@ ConnectionStateMachine(WorkerSession *session)
 
 					connection->waitFlags = WL_SOCKET_WRITEABLE;
 					connection->connectionState = MULTI_CONNECTION_CONNECTED;
-					break;
 				}
 
+				/* connection->waitFlags has changed, we need to rebuild the WaitEventSet */
+				execution->rebuildWaitEventSet = true;
```

Comment:
As long as we're not adding or removing (in case of a failure) a connection, we could use `ModifyWaitEvent` and avoid rebuilding the `WaitEventSet`

Reply:
done